src/mongocrypt.c

```diff
@@ -534,53 +551,28 @@ _try_replace_dollar_origin (mstr *filepath, _mongocrypt_log_t *log)
 _loaded_csfle
 _try_find_csfle (mongocrypt_t *crypt)
 {
-   _loaded_csfle candidate_csfle = {0};
-   mstr csfle_cand_filepath = MSTR_NULL;
+   _loaded_csfle candidate_csfle;
+   mstr csfle_search_name = MSTR_NULL;
    if (crypt->opts.csfle_lib_override_path.data) {
       // If an override path was specified, skip the library searching behavior
-      csfle_cand_filepath =
-         mstr_copy (crypt->opts.csfle_lib_override_path.view);
-      if (_try_replace_dollar_origin (&csfle_cand_filepath, &crypt->log)) {
+      csfle_search_name = mstr_copy (crypt->opts.csfle_lib_override_path.view);
+      if (_try_replace_dollar_origin (&csfle_search_name, &crypt->log)) {
          // Succesfully substituted $ORIGIN
          // Do not allow a plain filename to go through, as that will cause the
          // DLL load to search the system.
-         mstr_assign (&csfle_cand_filepath,
-                      mpath_absolute (csfle_cand_filepath.view, MPATH_NATIVE));
-         candidate_csfle =
-            _try_load_csfle (csfle_cand_filepath.data, &crypt->log);
+         mstr_assign (&csfle_search_name,
+                      mpath_absolute (csfle_search_name.view, MPATH_NATIVE));
       }
    } else {
-      // No override path was specified, so try to find it on the provided
-      // search paths.
-      for (int i = 0; i < crypt->opts.n_cselib_search_paths; ++i) {
-         mstr_view cand_dir = crypt->opts.cselib_search_paths[i].view;
-         mstr_view csfle_filename = mstrv_lit ("mongo_csfle_v1" MCR_DLL_SUFFIX);
-         if (mstr_eq (cand_dir, mstrv_lit ("$SYSTEM"))) {
-            // Caller wants us to search for the library on the system's default
-            // library paths. Pass only the library's filename to cause dll_open
-            // to search on the library paths.
-            mstr_assign (&csfle_cand_filepath, mstr_copy (csfle_filename));
-         } else {
-            // Compose the candidate filepath:
-            mstr_assign (&csfle_cand_filepath,
-                         mpath_join (cand_dir, csfle_filename, MPATH_NATIVE));
-            if (!_try_replace_dollar_origin (&csfle_cand_filepath,
-                                             &crypt->log)) {
-               // Error while substituting $ORIGIN
-               continue;
-            }
-         }
-         // Try to load the file:
-         candidate_csfle =
-            _try_load_csfle (csfle_cand_filepath.data, &crypt->log);
-         if (candidate_csfle.okay) {
-            // Stop searching:
-            break;
-         }
-      }
+      // No override path was specified, so try to find it on the system's
+      // default library paths. Pass only the library's filename to cause
+      // dll_open to search on the library paths.
+      csfle_search_name = mstr_copy_cstr ("mongo_csfle_v1");
```

Comment:
Are we testing this somehow? I don’t think this is going to work without a suffix on at least Linux, the dynamic loader doesn’t do extension searching itself:

```c
$ cat test.c
#include <dlfcn.h>
#include <stdio.h>

int main() {
  void* handle = dlopen("librt.so", RTLD_GLOBAL | RTLD_NOW);
  fprintf(stderr, "handle=%p, dlerror=%s\n", handle, dlerror());
  handle = dlopen("librt", RTLD_GLOBAL | RTLD_NOW);
  fprintf(stderr, "handle=%p, dlerror=%s\n", handle, dlerror());
}
$ gcc -o test test.c -ldl && ./test
handle=0x5579b495c2d0, dlerror=(null)
handle=(nil), dlerror=librt: cannot open shared object file: No such file or directory
```

(Also, this is more a tiny nit and something you can feel free to ignore, but wouldn’t `mstrv_view`/`mstrv_lit` work here now as well?)

Reply:
`mstr_view` won't work in this case, because in the path override case we do a string substitution in the path for `$ORIGIN`.

Good catch on the extension. This is difficult to test without environment trickery (see my later comment).